tests/Core/Tokenizer/BackfillEnumTest.php

```diff
@@ -0,0 +1,229 @@
+
+        $this->assertSame($expectedScopeOpener, $scopeOpener);
+        $this->assertArrayHasKey('scope_condition', $tokens[$scopeOpener]);
+        $this->assertArrayHasKey('scope_opener', $tokens[$scopeOpener]);
+        $this->assertArrayHasKey('scope_closer', $tokens[$scopeOpener]);
+        $this->assertSame($enum, $tokens[$scopeOpener]['scope_condition']);
+        $this->assertSame($scopeOpener, $tokens[$scopeOpener]['scope_opener']);
+        $this->assertSame($scopeCloser, $tokens[$scopeOpener]['scope_closer']);
+
+        $this->assertSame($expectedScopeCloser, $scopeCloser);
+        $this->assertArrayHasKey('scope_condition', $tokens[$scopeCloser]);
+        $this->assertArrayHasKey('scope_opener', $tokens[$scopeCloser]);
+        $this->assertArrayHasKey('scope_closer', $tokens[$scopeCloser]);
+        $this->assertSame($enum, $tokens[$scopeCloser]['scope_condition']);
+        $this->assertSame($scopeOpener, $tokens[$scopeCloser]['scope_opener']);
+        $this->assertSame($scopeCloser, $tokens[$scopeCloser]['scope_closer']);
+
+    }//end testEnums()
+
+
+    /**
+     * Data provider.
+     *
+     * @see testEnums()
+     *
+     * @return array
+     */
+    public function dataEnums()
+    {
+        return [
+            [
+                '/* testPureEnum */',
+                'enum',
+                4,
+                12,
+            ],
+            [
+                '/* testBackedIntEnum */',
+                'enum',
+                6,
+                28,
+            ],
+            [
+                '/* testBackedStringEnum */',
+                'enum',
+                6,
+                28,
+            ],
+            [
+                '/* testComplexEnum */',
+                'enum',
+                10,
+                71,
+            ],
+            [
+                '/* testEnumWithEnumAsClassName */',
+                'enum',
+                6,
+                7,
+            ],
+            [
+                '/* testEnumIsCaseInsensitive */',
+                'EnUm',
+                4,
+                5,
+            ],
+            [
+                '/* testDeclarationContainingComment */',
+                'enum',
+                6,
+                14,
+            ],
+        ];
+
+    }//end dataEnums()
+
+
+    /**
+     * Test that "enum" when not used as the keyword is still tokenized as `T_STRING`.
+     *
+     * @param string $testMarker  The comment which prefaces the target token in the test file.
+     * @param string $testContent The token content to look for.
+     *
+     * @dataProvider dataNotEnums
+     * @covers       PHP_CodeSniffer\Tokenizers\PHP::tokenize
+     *
+     * @return void
+     */
+    public function testNotEnums($testMarker, $testContent)
+    {
+        $tokens = self::$phpcsFile->getTokens();
+
+        $target = $this->getTargetToken($testMarker, [T_ENUM, T_STRING], $testContent);
+        $this->assertSame(T_STRING, $tokens[$target]['code']);
```

Comment:
Due to the way namespaces are tokenized in 4.0 (they become `T_NAME_RELATIVE` or `T_NAME_QUALIFIED`) I had to reverse this logic in the 4.0 branch to use `assertNotSame(T_ENUM...)`. I didn't make the same change in master.